### PR TITLE
Editor: Introduce WP_Block and WP_Block_List classes

### PR DESCRIPTION
**Pull request by gziolo**
Trac ticket: https://core.trac.wordpress.org/ticket/49926

This PR ports the most up-to-date version of the `WP_Block` and `WP_Block_List` classes added to Gutenberg:

https://github.com/WordPress/gutenberg/blob/master/lib/class-wp-block.php
https://github.com/WordPress/gutenberg/blob/master/lib/class-wp-block-list.php

It also adds corresponding unit tests.

There is a follow-up backport necessary to wire this class when using new block context features as tracked in https://core.trac.wordpress.org/ticket/49927.

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-includes/class-wp-block-list.php [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+
+/**
+ * Class representing a list of block instances.
+ */
+class WP_Block_List implements Iterator, ArrayAccess, Countable {
+
+	/**
+	 * Original array of parsed block data.
+	 *
+	 * @var array|WP_Block[]
```
Comment (TimothyBJacobs) [inline src/wp-includes/class-wp-block-list.php:17]: The `array` part of the type isn't necessary.

File: src/wp-includes/class-wp-block-list.php [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+	 * Constructor.
+	 *
+	 * Populates object properties from the provided block instance argument.
+	 *
+	 * @param array|WP_Block[]       $blocks            Array of parsed block data, or block instances.
+	 * @param array                  $available_context Optional array of ancestry context values.
+	 * @param WP_Block_Type_Registry $registry          Optional block type registry.
+	 */
+	public function __construct( $blocks, $available_context = array(), $registry = null ) {
+		if ( is_null( $registry ) ) {
```
Comment (TimothyBJacobs) [inline src/wp-includes/class-wp-block-list.php:48]:
```suggestion
		if ( ! $registry instanced WP_Block_Type_Registry ) {
```

Would be more type safe.

File: src/wp-includes/class-wp-block-list.php [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+			$registry = WP_Block_Type_Registry::get_instance();
+		}
+
+		$this->blocks            = $blocks;
+		$this->available_context = $available_context;
+		$this->registry          = $registry;
+	}
+
+	/*
+	 * ArrayAccess interface methods.
```
Comment (TimothyBJacobs) [inline src/wp-includes/class-wp-block-list.php:58]: I don't think these kinds of comments are typically used, see `WP_Hook` for instance. I don't know if we have any coding standards around this.

File: src/wp-includes/class-wp-block.php [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+	 * @param array $options {
+	 *   Optional options object.
+	 *
+	 *   @type bool $dynamic Defaults to 'true'. Optionally set to false to avoid using the block's render_callback.
+	 * }
+	 * @return string Rendered block output.
+	 */
+	public function render( $options = array() ) {
+		global $post;
+		$options = array_replace(
```
Comment (TimothyBJacobs) [inline src/wp-includes/class-wp-block.php:181]: Why not `wp_parse_args` here?

**Comment by TimothyBJacobs** [comment]
I think this needs `@since` comments on all the classes, methods, and properties not just for the file.

File: src/wp-includes/class-wp-block-list.php (hunk 1 above)
Comment (gziolo) [reply src/wp-includes/class-wp-block-list.php:17]: I think it's correct as is, see line 43:

`* @param array|WP_Block[]       $blocks            Array of parsed block data, or block instances.`

File: src/wp-includes/class-wp-block-list.php (hunk 3 above)
Comment (gziolo) [reply src/wp-includes/class-wp-block-list.php]: Removed.

File: src/wp-includes/class-wp-block.php (hunk 4 above)
Comment (gziolo) [reply src/wp-includes/class-wp-block.php:181]: I don't know but I can update code :)

Comment (gziolo) [reply src/wp-includes/class-wp-block.php]: Updated in 7385d10.

File: src/wp-includes/class-wp-block-list.php (hunk 1 above)
Comment (TimothyBJacobs) [reply src/wp-includes/class-wp-block-list.php:20]: Ah maybe we can include that in the property description? Also perhaps expand to be `array[]|WP_Block[]`? 

To me it reads at first glance of saying this is an array and its an array of `WP_Block` instances. I think those changes might make that more clear?

Comment (gziolo) [reply src/wp-includes/class-wp-block-list.php]: Good call, I checked the code (I didn't author it) and it looks like what you proposed makes the most sense 👍 Updated.

**Pull request closed without merging**